### ISSUE TITLE
Microsoft.CodeAnalysis.CSharp.CodeStyle fails at runtime 

### ISSUE THREAD

**Issue by ViktorHofer** [dotnet/roslyn#50833]
**Version Used**: Microsoft.CodeAnalysis.CSharp.CodeStyle/3.9.0-3.final

**Steps to Reproduce**: Package reference `Microsoft.CodeAnalysis.CSharp.CodeStyle/3.9.0-3.final` and add rule that triggers `UseExpressionBodyDiagnosticAnalyzer` be instantiated.

**Expected Behavior**: No error at run time and dependency to Microsoft.CodeAnalysis expressed correctly.

**Actual Behavior**:
```
CSC : error CS8032: An instance of analyzer Microsoft.CodeAnalysis.CSharp.UseExpressionBody.UseExpressionBodyDiagnosticAnalyzer cannot be created from C:\Users\vihofer\.nuget\packages\microsoft.codeanalysis.csharp.codestyle\3.9.0-3.final\analyzers\dotnet\cs\Microsoft.CodeAnalysis.CSharp.CodeStyle.dll : Could not load file or assembly 'Microsoft.CodeAnalysis, Version=3.9.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'. The system cannot find the file specified.. [C:\git\runtime2\src\libraries\System.Resources.Reader\src\System.Resources.Reader.csproj]
```

My assumption is that `Microsoft.CodeAnalysis.CSharp.CodeStyle` requires Microsoft.CodeAnalysis.dll which is shipped via the Microsoft.CodeAnalysis.Common package which would mean that the CodeStyle package should have add a package dependency. Because of https://github.com/dotnet/roslyn/issues/50832, the Common package is currently not usable in dotnet/runtime libraries.

cc @sharwell @stephentoub

**Comment by sharwell** [comment]
The CodeStyle package requires a matching or newer version of the compiler. In other words, version 3.9.0-3 requires the following:

* Command line use requires Roslyn 3.9 Preview 3 or newer
* IDE use requires Visual Studio 2019 version 16.9 Preview 3 or newer

Can you verify that the repository is not using a version older than supported by the analyzers?

**Comment by ViktorHofer** [comment]
We are using the version that is pinned by Arcade: https://github.com/dotnet/arcade/blob/afd54cd0cb49f8c919411394231b1c7447513a51/eng/Versions.props#L38.

> The CodeStyle package requires a matching or newer version of the compiler.

Would it make sense to indicate the requirement of a matching compiler in the package description?

Just for my curiosity, can you explain the relationship of Microsoft.CodeAnalysis.dll and a newer compiler version?

**Comment by sharwell** [comment]
> We are using the version that is pinned by Arcade

This is also the newest allowed version of the CodeStyle layer. Do keep in mind that the CodeStyle layer has impact on both command line and IDE scenarios, so in some cases you might want to intentionally hold back the CodeStyle version when updating the compiler to avoid requiring developers on the project to update the IDE.

> Would it make sense to indicate the requirement of a matching compiler in the package description?

Seems yes.

> Just for my curiosity, can you please explain the relationship of Microsoft.CodeAnalysis.dll and a newer compiler version?

Microsoft.CodeAnalysis is provided by either the compiler or by the IDE, depending on what you're currently doing. The CodeStyle layer requires the application loading the CodeStyle layer provide a version at or newer than the CodeStyle layer.

**Comment by ViktorHofer** [comment]
> Microsoft.CodeAnalysis is provided by either the compiler or by the IDE, depending on what you're currently doing. The CodeStyle layer requires the application loading the CodeStyle layer provide a version at or newer than the CodeStyle layer. If this isn't clear let me know since I may have misinterpreted the question.

What is the Microsoft.CodeAnalysis.Common package for then which includes Microsoft.CodeAnalysis.dll?

**Comment by sharwell** [comment]
> What is the Microsoft.CodeAnalysis.Common package for then which includes Microsoft.CodeAnalysis.dll?

If you want to write code that uses the compiler APIs (e.g. for an analyzer, source generator, or standalone tool), you would need to reference this package.
